Library/Homebrew/livecheck/livecheck.rb

```diff
@@ -0,0 +1,404 @@
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module_function
+
+    GITHUB_SPECIAL_CASES = %w[
+      api.github.com
+      /latest
+      mednafen
+      camlp5
+      kotlin
+      osrm-backend
+      prometheus
+      pyenv-virtualenv
+      sysdig
+      shairport-sync
+      yuicompressor
+    ].freeze
+
+    UNSTABLE_VERSION_KEYWORDS = %w[
+      alpha
+      beta
+      bpo
+      dev
+      experimental
+      prerelease
+      preview
+      rc
+    ].freeze
+
+    def livecheck_formulae(formulae_to_check, args)
+      # Identify any non-homebrew/core taps in use for current formulae
+      non_core_taps = {}
+      formulae_to_check.each do |f|
+        next if f.tap.blank?
+        next if f.tap.name == CoreTap.instance.name
+        next if non_core_taps[f.tap.name]
+
+        non_core_taps[f.tap.name] = f.tap
+      end
+      non_core_taps = non_core_taps.sort.to_h
+
+      # Load additional Strategy files from taps
+      non_core_taps.each_value do |tap|
+        tap_strategy_path = "#{tap.path}/livecheck/strategy"
+        Dir["#{tap_strategy_path}/*.rb"].sort.each(&method(:require)) if Dir.exist?(tap_strategy_path)
+      end
+
+      # Cache demodulized strategy names, to avoid repeating this work
+      @livecheck_strategy_names = {}
+      Strategy.constants.sort.each do |strategy_symbol|
+        strategy = Strategy.const_get(strategy_symbol)
+        @livecheck_strategy_names[strategy] = strategy.name.demodulize
+      end
+      @livecheck_strategy_names.freeze
+
+      has_a_newer_upstream_version = false
+      formulae_checked = formulae_to_check.sort.map.with_index do |formula, i|
+        if args.debug? && i.positive?
+          puts <<~EOS
+
+            ----------
+
+          EOS
+        end
+
+        skip_result = skip_conditions(formula, args: args)
+        next skip_result if skip_result != false
+
+        formula.head.downloader.shutup! unless formula.stable?
+
+        current = formula.stable? ? formula.version : formula.installed_version.version.commit
+
+        version_info = nil
```

Comment:
I think with modern Ruby you may not need this line but I'll let you figure out either way.

Reply:
It seems like it may still be necessary for head-only formulae. `version_info` wouldn't be set for a HEAD-only formula but it's referenced a few lines later in:

```ruby
if latest.nil?
  if args.json?
    next version_info if version_info.is_a?(Hash) && version_info["status"] && version_info["messages"]

    next status_hash(formula, "error", [NO_VERSIONS_MSG], args: args)
  else
    raise TypeError, NO_VERSIONS_MSG
  end
end
```

If `version_info` isn't defined, `version_info.is_a?(Hash)` would seemingly give us a ```NameError (undefined local variable or method `version_info' for main:Object)``` error.

Reply:
> If `version_info` isn't defined, `version_info.is_a?(Hash)` would seemingly give us a `` NameError (undefined local variable or method `version_info' for main:Object) `` error.

If you've tested this: 👍🏻 with me. It doesn't seem to always be the case in Ruby now.

Reply:
Just to be sure, I commented out the `version_info = nil` line and tested this in `brew livecheck` with a HEAD-only formula (which I put together just to test this). In this context, `version_info` is still set to `nil` even when we don't do it manually, so it's fine to remove the `version_info = nil` line.

I added some output statements in various places to confirm that this was going down the relevant code path (not doing something else) and it all checked out.

I also tested this in a standalone script with some similar code and it seems Ruby handles the `nil` due to `version_info` being set in a if/else statement. If you remove the if/else code, then we get the expected `undefined local variable` error.

I wasn't aware of this before, so thanks for the heads up.

Reply:
> In this context, `version_info` is still set to `nil` even when we don't do it manually, so it's fine to remove the `version_info = nil` line.

Yup, that's what I thought, thanks for checking.



> I wasn't aware of this behavior before, so thanks for the heads up.

It's newish? I can't remember what Ruby version it landed in, I think one between 2.3 and 2.6?